Update workshop draft filter validation logic and Moderator/TechAdmin authorization

max-khakimzhanov opened the pull request:
This PR enhances the workshop draft filtering system by updating validation logic and strengthening moderation checks to improve content quality and data integrity. Additionally, it introduces comprehensive authorization checks for Moderator and TechAdmin roles to ensure users exist in the corresponding database tables before granting access.

## Changes Made

### Enhanced Authorization System
**Updated CurrentUserService with robust role-based security:**
- **Added new dependencies**: `ISensitiveEntityRepositorySoftDeleted<Moderator>` and `ISensitiveEntityRepositorySoftDeleted<TechAdmin>` repositories
- **Updated private methods**: `ModeratorHasRights()` and `TechAdminHasRights()` for database validation
- **Updated UserHasRights logic**: Added support for `ModeratorRights` and `TechAdminRights` in the authorization pipeline
- **Enhanced security**: Users are now automatically logged out if removed from Moderator/TechAdmin tables, even before token expiration

**New Authorization Flow:**
1. **Role validation**: Check if user has the required role claim
2. **Database validation**: Verify user exists in the corresponding table via `Individual.UserId` lookup
3. **Audit logging**: Log unauthorized access attempts when `AccessLogEnabled` is true
4. **Graceful handling**: Return false for missing users without exposing system details

### Workshop Draft Filter Validation
- **Updated validation rules** for workshop draft filtering

### Moderation System Improvements
- **Enhanced content moderation checks** for workshop drafts before publication

### Added tests for `ModeratorHasRights()` and `TechAdminHasRights()`

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Moderation actions on workshop drafts now support multiple statuses, allowing for broader moderation scenarios.
- **Refactor**
  - Moderator and tech admin permissions are now determined by user context rather than requiring explicit user IDs.
  - Workshop draft moderation endpoints and related services have simplified method signatures, no longer requiring moderator IDs.
- **Bug Fixes**
  - Enhanced validation for draft statuses, ensuring only allowed statuses are processed during moderation.
- **Tests**
  - Updated and added tests to reflect changes in permission checks and method signatures for moderation actions.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

max-khakimzhanov committed:
Update WorkshopDraftStatusValidationAttribute and moderation checks

Updated `WorkshopDraftService` to include additional checks
for `DraftStatus` during approval and rejection, ensuring
that drafts edited by moderators cannot be approved or
rejected. Also modified `AllowedStatuses` in
`WorkshopDraftStatusValidationAttribute` to include
`WorkshopDraftStatus.EditedByModerator` for consistency.

max-khakimzhanov committed:
Update parameters and documentation of WorkshopDraft editing methods

Updated `ISensitiveWorkshopDraftService` to replace `moderatorId` with `userId` in multiple methods. Adjusted `WorkshopDraftService` and `WorkshopDraftController` to reflect these changes, ensuring clarity that actions. Documentation has been updated accordingly.

max-khakimzhanov committed:
Refactor WorkshopDraft status handling and validation

- Updated `WorkshopDraftFilterAdministration` to use a
  `HashSet<WorkshopDraftStatus>` for multiple statuses and
  added a method for default statuses.

- Modified filtering logic in `WorkshopDraftService` to
  accommodate the new collection of statuses.

- Enhanced `WorkshopDraftStatusValidationAttribute` to
  validate collections of statuses, providing detailed
  error messages for invalid entries.

max-khakimzhanov committed:
Update TechAdmin and Moderator rights validation and service interfaces

- Updated `CurrentUserService` to include new dependencies for moderator and tech admin repositories, removing user ID requirements for rights checks.
- Simplified `ISensitiveWorkshopDraftService` by removing `userId` parameters from several methods.
- Adjusted `WorkshopDraftService` to align with the updated interface, streamlining user rights validation logic.
- Modified `WorkshopDraftController` to reflect changes in method signatures, eliminating unnecessary `userId` parameters.
- Revised `TechAdminRights` and `ModeratorRights` records to no longer require an ID for validation.

max-khakimzhanov committed:
Refactor tests to remove moderatorId parameter

Removed the `moderatorId` parameter from method calls in
`WorkshopDraftControllerTests` and `SensitiveWorkshopDraftServiceTests`
to simplify method signatures. Updated tests to ensure functionality
remains intact.

Added a new `CurrentUserServiceTests` class with tests to verify
the rights of moderators and tech admins, utilizing mocking for
dependencies. Tests now assert correct behavior based on user roles
and access logging settings, enhancing clarity and maintainability.

max-khakimzhanov committed:
Remove userId parameter from ValidateDraftForModerator

DmyMi commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/CurrentUserService.cs:
```diff
@@ -295,43 +297,38 @@ private async Task<bool> EmployeeWorkshopRights(Guid workshopId)
         }
 
-        var result = UserId == moderatorId.ToString();
+        var result = await moderatorRepository.Any(m => m.Individual.UserId == UserId);
```
Technically, moderator/techadmin `Id` is both a primary key to moderator/techadmin as well as foreign key to individual.
As this is not the issue now, but we can add individual `Id` to token during auth and just do the `Any(m => m.Id = <claim_from_toekn>)`

DmyMi commented on OutOfSchool/OutOfSchool.BusinessLogic/Models/WorkshopDraft/WorkshopDraftFilterAdministration.cs:
```diff
@@ -1,12 +1,25 @@
 namespace OutOfSchool.BusinessLogic.Models.WorkshopDraft;
 public class WorkshopDraftFilterAdministration : WorkshopFilterAdministration
 {
-    [EnumDataType(typeof(WorkshopDraftStatus), ErrorMessage = Constants.EnumErrorMessage)]
```
need to think about how to do this check for enum datatype, because it is possible in some cases to pass an array of random integers and this will deserialize into broken enums.

max-khakimzhanov committed:
Update authorization checks in GetWorkshopDraftByIdMapped

Modified the `GetWorkshopDraftByIdMapped` method in the `WorkshopDraftService` class. The method now verifies user rights for `ProviderRights`, `EmployeeRights`, `ModeratorRights`, and `TechAdminRights`, broadening access to workshop draft details.
